File: 백준/Bronze/9506. 약수들의 합/약수들의 합.py
def divisor(n):
    divlist = []
    for i in range(1, int(n**(1/2)+1)):
        if n%i==0:
            divlist.append(i)
            # i**2==n이면 중복으로 두개가 들어갈 수 있으므로
            if i**2!=n:
                divlist.append(n//i)
    divlist.sort()
    return divlist

while True:
    n = int(input())
    if n==-1:
        break
    divlist = divisor(n)
    if sum(divlist) == 2 * n:
        print(n, end=' = ')
        for i in range(len(divlist)-2):
            print(divlist[i], end=' + ')
        # 마지막에서 두번째는 따로 출력
        print(divlist[-2])
    else:
        print(n, "is NOT perfect.")

File: 백준/Bronze/9506. 약수들의 합/약수들의 합.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

// 제곱근을 사용하여 약수들을 리턴하는 함수
vector<int> getDivisors(int n) {
    vector<int> divisors;

    int i=1;
    while (i * i < n){
        if (n % i == 0) {
            divisors.push_back(i);
            divisors.push_back(n / i);
        }
        i++;
    }
    // 제곱수인경우 중복하지 않도록 하나만 추가
    if (i*i == n) divisors.push_back(i);
    
    return divisors;
}

int main() {
    while (true){
        int N;
        cin >> N;
        if (N==-1) break;
        vector<int> divisors = getDivisors(N);
        sort(divisors.begin(), divisors.end());
        divisors.pop_back();

        // 벡터 요소 합 accumulate
        int sum = accumulate(divisors.begin(), divisors.end(), 0);
        if (sum == N){
            cout << N << " = ";

            for (int i=0; i<divisors.size()-1; i++){
                cout << divisors[i] << " + ";
            }
            cout << divisors[divisors.size()-1] << '\n';
        }
        else cout << N << " is NOT perfect.\n";
    }

    return 0;
}
